lib/client/api.go

```diff
@@ -504,28 +500,22 @@ func VirtualPathEnvNames(kind VirtualPathKind, params VirtualPathParams) []strin
 // RetryWithRelogin is a helper error handling method, attempts to relogin and
 // retry the function once.
 func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error) error {
-	err := fn()
-	if err == nil {
+	fnErr := fn()
+	switch {
+	case fnErr == nil:
 		return nil
+	case tc.NonInteractive:
+		return trace.Wrap(fnErr)
+	case utils.IsPredicateError(fnErr):
+		return trace.Wrap(utils.PredicateError{Err: fnErr})
+	case !IsErrorResolvableWithRelogin(fnErr):
+		return trace.Wrap(fnErr)
```

Comment:
Should we push IsPredicateError up, so that handling happens regardless of mode?

```suggestion
	case utils.IsPredicateError(fnErr):
		return trace.Wrap(utils.PredicateError{Err: fnErr})
	case tc.NonInteractive, !IsErrorResolvableWithRelogin(fnErr):
		return trace.Wrap(fnErr)
```